design/001-secretsink.md

```diff
@@ -0,0 +1,188 @@
+
+### Goals
+- CRD Design for the SecretSink
+- Define the need for a SinkStore
+- 
+### Non-Goals
+Do not implement full compatibility mechanisms with each provider (we are not Terraform neither Crossplane)
+
+### Terminology
+- Sink object: any Secret (a part or the whole secret) from Kubernetes that is going to be uploaded to a Provider.
+## Proposal
+
+A controller that checks for Sink Objects, gets K8s Secrets and creates the equivalent secret on the SecretStore Provider.
+
+### User Stories
+1. As an ESO Operator I want to be able to Sync Secrets in my cluster with my External Provider
+1. As an ESO Operator I want to be able to Sync Secrets even if they are not bound to a given ExternalSecret
+
+### API
+Proposed CRD changes:
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: SecretStore
+metadata:
+  name: example
+  namespace: example-ns
+spec:
+  controller: dev
+  retrySettings:
+    maxRetries: 5
+    retryInterval: "10s"
+  provider:
+    aws:
+      service: SecretsManager
+      role: iam-role
+      region: eu-central-1
+      encryptionConfig: {} # Specific config for Creating Secrets on AWS
+      auth:
+        secretRef:
+          accessKeyID:
+            name: awssm-secret
+            key: access-key
+          secretAccessKey:
+            name: awssm-secret
+            key: secret-access-key
+    vault:
+      server: "https://vault.acme.org"
+      path: "secret"
+      version: "v2"
+      encryptionConfig: {} # Specific config for Creating Secrets on Vault ()
+      namespace: "a-team"
+      caBundle: "..."
+      caProvider:
+        type: "Secret"
+        name: "my-cert-secret"
+        key: "cert-key"
+      auth:
+        tokenSecretRef:
+          name: "my-secret"
+          namespace: "secret-admin"
+          key: "vault-token"
+        appRole:
+          path: "approle"
+          roleId: "db02de05-fa39-4855-059b-67221c5c2f63"
+          secretRef:
+            name: "my-secret"
+            namespace: "secret-admin"
+            key: "vault-token"
+        kubernetes:
+          mountPath: "kubernetes"
+          role: "demo"
+          serviceAccountRef:
+            name: "my-sa"
+            namespace: "secret-admin"
+          secretRef:
+            name: "my-secret"
+            namespace: "secret-admin"
+            key: "vault"
+    gcpsm:
+      encryptionConfig: {} # Specific config for Creating Secrets GCP SM
+      auth:
+        secretRef:
+          secretAccessKeySecretRef:
+            name: gcpsm-secret
+            key: secret-access-credentials
+      projectID: myproject
+```
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: SecretSink
+metadata:
+  name: "hello-world"
+  namespace: my-ns # Same of the SecretStores
+spec:
+  secretStoreRefs:
+  - name: secret-store
+    kind: SecretStore
+  - name: secret-store-2
+    kind: SecretStore
+  - name: cluster-secret-store
+    kind: ClusterSecretStore
+  refreshInterval: "1h"
+
+  data:
```

Comment:
Do we really need the `.data` node here? I think `.selector` doesn't really fit here and could be hoisted up to `spec.selector`.
IMO the same goes for `.match` and `.rewrite` - they could also be hoisted up to `.spec`.

Reply:
The idea here is to allow more features to come in afterwards and already be 'encapsulated' nicely (for instance in GCP SM it is possible to set replication policy, expiration and labels on a secret level)

So the data here was to try to encapsulate that this information is regarding the data that goes in the provider Secret, not any other configuration.

I agree with you for the selector bit of it. Maybe we can remove the `data` field if we are sure that this secret 'configuration' is going to be stored in the store level. I just didn't want to commit to that yet